pkg/kubelet/nodeconfig/parse.go

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodeconfig
+
+import (
+	"encoding/json"
+	"fmt"
+
+	yaml "k8s.io/apimachinery/pkg/util/yaml"
+	api "k8s.io/kubernetes/pkg/api"
+	apiv1 "k8s.io/kubernetes/pkg/api/v1"
+	ccv1a1 "k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1"
+)
+
+// parsable is for parsing a KubeletConfiguration out of a config object
+type parsable interface {
+	parse() (*ccv1a1.KubeletConfiguration, error)
+}
+
+// parsableConfigMap is for parsing a KubeletConfiguration out of a ConfigMap
+type parsableConfigMap struct {
+	cm *apiv1.ConfigMap
+}
+
+// parse deserializes the KubeletConfiguration out of the ConfigMap's Data, and also applies defaults.
+// If parsing fails, returns an error.
+// If parsing succeeds, returns a KubeletConfiguration.
+// It is recommended that you validate any returned configuration before using it.
+func (p *parsableConfigMap) parse() (*ccv1a1.KubeletConfiguration, error) {
+	if len(p.cm.Data) == 0 {
+		return nil, fmt.Errorf("configuration was empty, but some parameters are required")
+	}
+
+	// TODO(mtaufen): Once the KubeletConfiguration type is decomposed (#44252), allow multiple keys to contain configuration.
+	// Since we presently only expect one key, the for loop is a convenient way to express "take any value in the map".
```

Comment:
In general, it wouldn't be safe to roll out newly-typed configuration until all clients are updated to understand it. If we were planning on adding a new config object, we'd have to make it optional, upgrade the Kubelets to understand it, and then roll out the new configurations that use that object.

That said, I see some usefulness to being able to express a transition of configuration across Kubelet upgrades in the ConfigMap. What would you think of requiring the key name match the minimum version of the Kubelet that should be allowed to use a given configuration, e.g. `1.7.0`, or a key expressing the version of the API group that the config came from, e.g. `kubelet.v1alpha1`? A tool could definitely be created to understand these relationships.

Alternatively, we could just glue each config object to a key name, and Kubelets could just ignore keys they don't understand.

For this PR, I'll just require the config live in a key called `kubelet.v1alpha1`, and we can keep discussing this as the feature develops.